metadata/src/audio/item.rs

```diff
@@ -16,98 +18,290 @@ use librespot_core::{
-        user_data: &UserData,
-        restrictions: &Restrictions,
-    ) -> AudioItemAvailability {
-        let country = &user_data.country;
-        let user_catalogue = match user_data.attributes.get("catalogue") {
-            Some(catalogue) => catalogue,
-            None => "premium",
-        };
-
-        for premium_restriction in restrictions.iter().filter(|restriction| {
-            restriction
-                .catalogue_strs
-                .iter()
-                .any(|restricted_catalogue| restricted_catalogue == user_catalogue)
-        }) {
-            if let Some(allowed_countries) = &premium_restriction.countries_allowed {
-                // A restriction will specify either a whitelast *or* a blacklist,
-                // but not both. So restrict availability if there is a whitelist
-                // and the country isn't on it.
-                if allowed_countries.iter().any(|allowed| country == allowed) {
-                    return Ok(());
-                } else {
-                    return Err(UnavailabilityReason::NotWhitelisted);
+                if track.duration <= 0 {
+                    return Err(Error::unavailable(MetadataError::InvalidDuration(
+                        track.duration,
+                    )));
                 }
-            }
 
-            if let Some(forbidden_countries) = &premium_restriction.countries_forbidden {
-                if forbidden_countries
+                if track.is_explicit {
+                    if let Some(value) = session.get_user_attribute("filter-explicit-content") {
+                        if &value == "1" {
+                            return Err(Error::unavailable(MetadataError::ExplicitContentFiltered));
+                        }
+                    }
+                }
+
+                let track_id = track.id;
+                let uri = track_id.to_uri()?;
+                let album = track.album.name;
+
+                let artists = track
+                    .artists
+                    .iter()
+                    .map(|a| a.name.clone())
+                    .collect::<Vec<String>>();
+
+                let album_artists = track
+                    .album
+                    .artists
                     .iter()
-                    .any(|forbidden| country == forbidden)
-                {
-                    return Err(UnavailabilityReason::Blacklisted);
+                    .map(|a| a.name.clone())
+                    .collect::<Vec<String>>();
+
+                let mut covers = track.album.covers;
+
+                covers.sort_by(|a, b| b.width.cmp(&a.width));
+
+                let covers = covers
+                    .iter()
+                    .filter_map(|cover| {
+                        let cover_id = cover.id.to_string();
+
+                        if !cover_id.is_empty() {
+                            let cover_image = CoverImage {
+                                url: image_url.clone().replace("{file_id}", &cover_id),
```

Comment:
```suggestion
                                url: image_url.replace("{file_id}", &cover_id),
```

[`str::replace()`](https://doc.rust-lang.org/std/primitive.str.html#method.replace) allocates already itself, so no need to clone here.

Reply:
Suggest removing the other clone and I will commit it. I'm on my phone right now so that's the easiest way to get that changed.